File: src/core/chain.py
import json
from .block import Block

class Chain:
  def __init__(self, name='', difficulty={'char': '0', 'times': 4}):
    self.name = name
    self.difficulty = difficulty
    self.blocks = [self.createGenesis()]

  def to_dict(self):
    return {
      'name': self.name,
      'blocks': list(map(lambda x: x.to_dict(), self.blocks)),
    }
  
  def __str__(self):
    return json.dumps(self.to_dict())

  def createGenesis(self):
    return Block(0, '', 'genesis')

  def add(self, data={}):
    index = len(self.blocks)
    prev_hash = self.blocks[index - 1].meta.hash
    block = Block(index, prev_hash, data)
    block.proof_of_work(self.difficulty)
    self.blocks.append(block)
    return block

  def peek(self):
    return self.blocks[len(self.blocks) - 1]

  def is_valid(self):
    for index, block in enumerate(self.blocks):
      before = self.blocks[index - 1] if index > 0 else None
      if block.meta.hash != block.create_hash(block.meta.nonce): return False
      if before != None and block.meta.prev_hash != before.meta.hash: return False

    return True

File: src/core/chain/chain.go
package chain

import (
	"bacot/src/core/block"
)

type Chain struct {
	Name string
	Difficulty int
	Blocks []*block.Block
}

func New(name string, difficulty int) *Chain {
	var blocks []*block.Block
	genesis := CreateGenesis()
	blocks = append(blocks, genesis)
	chain := &Chain{
		Name: name,
		Difficulty: difficulty,
		Blocks: blocks,
	}

	return chain
}

func CreateGenesis() *block.Block {
	genesis := block.New(0, "", "genesis")
	return genesis
}

func (chain *Chain) Add(data interface{}) *block.Block {
	length := len(chain.Blocks)
	prevHash := chain.Blocks[length-1].Meta.Hash
	newblock := block.New(length-1, prevHash, data)
	newblock.ProofOfWork(chain.Difficulty)
	chain.Blocks = append(chain.Blocks, newblock)
	return newblock
}

func (chain *Chain) Peek() *block.Block {
	return chain.Blocks[len(chain.Blocks) - 1]
}

func (chain *Chain) IsValid() bool {
	for i, currentblock := range chain.Blocks {
		var before *block.Block
		if i > 0 {
			before = chain.Blocks[i - 1]
		}

		if before != nil && currentblock.Meta.PrevHash != before.Meta.Hash {
			return false
		}

	}

	return true
}
